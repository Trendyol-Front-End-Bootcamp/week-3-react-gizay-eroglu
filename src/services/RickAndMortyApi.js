import axios from 'axios';

export const getCharacters = async (params = {}) => {
  try {
    const response = await axios.get("https://rickandmortyapi.com/api/character", { params });
  
    return response.data.results || [];  
  } catch (ex) {
    return [];
  } 
}

export const getCharacterById = async (id) => {
  try {
    const response = await axios.get(`https://rickandmortyapi.com/api/character/${id}`);
    
    if(response.data.error) {
      return null; // Character Not Found  
    }

    const character = response.data;
    character.lastFiveEpisodes = [];

    let count = 0;

    while (count < 5) {
      const episodeUrl = character.episode[count];

      if (!episodeUrl) {
        break;
      }

      const episodeResponse = await axios.get(episodeUrl);
      character.lastFiveEpisodes.push(episodeResponse.data);

      count++;
    }

    return character;
  } catch (ex) {
    return null;
  }
}