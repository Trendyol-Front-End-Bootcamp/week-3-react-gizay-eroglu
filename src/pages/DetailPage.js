import { useState, useEffect } from "react";
import { useHistory } from "react-router-dom";

import Header from "../components/Header";
import CardDetail from "../components/CardDetail";

import { getCharacterById } from "../services/RickAndMortyApi";

function DetailPage({ match }) {
  const {
    params: { id },
  } = match;

  const history = useHistory();
  const [character, setCharacter] = useState(null);

  const getCharacter = async () => {
    const character = await getCharacterById(id);
    setCharacter(character);
  };

  useEffect(() => {
    getCharacter();
  }, [id]);

  const handleGoHome = () => {
    history.push("/");
  };

  if (!character) {
    return <div className='card-loading'>Loading</div>;
  }

  return (
    <>
      <Header />
      <button className='go-homepage-button' onClick={handleGoHome}>
        Go Home
      </button>
      <CardDetail
        name={character.name}
        gender={character.gender}
        locationName={character.location.name}
        image={character.image}
        species={character.species}
        status={character.status}
        lastFiveEpisodes={character.lastFiveEpisodes}
      />
    </>
  );
}

export default DetailPage;
