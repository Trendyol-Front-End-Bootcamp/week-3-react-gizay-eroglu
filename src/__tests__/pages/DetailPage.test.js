import React from "react";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import { shallow, configure } from "enzyme";
import { getCharacterById } from "../../services/RickAndMortyApi";

import DetailPage from "../../pages/DetailPage";

configure({adapter: new Adapter()});

describe("Detail page test", function () {

  let wrapper;

  beforeEach(() => { 
    wrapper = shallow(<DetailPage match = {{ params: { id: 1 }}} />);
  })

  it("should render cardDetail", () => {
    expect(wrapper).not.toBeNull();
    expect(wrapper).toMatchSnapshot();
  })
})