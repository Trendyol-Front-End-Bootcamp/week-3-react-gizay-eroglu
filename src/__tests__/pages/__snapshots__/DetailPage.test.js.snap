// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Detail page test should render cardDetail 1`] = `ShallowWrapper {}`;
